modules/smartclipBidAdapter.js

```diff
@@ -0,0 +1,359 @@
+      var contextcustom = utils.deepAccess(bid, 'mediaTypes.video.context');
+      var placement = 1;
+      if (contextcustom == 'outstream') {
+        placement = 3;
+      }
+      let smartxReq = {
+        id: bid.bidId,
+        secure: secure,
+        video: {
+          w: contentWidth,
+          h: contentHeight,
+          mimes: mimes,
+          linearity: linearity,
+          minduration: minduration,
+          maxduration: maxduration,
+          startdelay: startdelay,
+          protocols: protocols,
+          minbitrate: minbitrate,
+          maxbitrate: maxbitrate,
+          delivery: delivery,
+          pos: pos,
+          placement: placement,
+          api: api,
+          ext: ext
+        },
+        tagid: tagId,
+        ext: {
+          'smart.bidpricetype': 1
+        }
+      };
+      if (utils.getBidIdParameter('bidfloor', bid.params) != '') {
+        smartxReq.bidfloor = utils.getBidIdParameter('bidfloor', bid.params);
+      }
+      if (utils.getBidIdParameter('bidfloorcur', bid.params) != '') {
+        smartxReq.bidfloorcur = utils.getBidIdParameter('bidfloorcur', bid.params);
+      }
+      if (bid.crumbs && bid.crumbs.pubcid) {
+        pubcid = bid.crumbs.pubcid;
+      }
+      const language = navigator.language ? 'language' : 'userLanguage';
+      const device = {
+        h: screen.height,
+        w: screen.width,
+        dnt: utils.getDNT() ? 1 : 0,
+        language: navigator[language].split('-')[0],
+        make: navigator.vendor ? navigator.vendor : '',
+        ua: navigator.userAgent
+      };
+      const at = utils.getBidIdParameter('at', bid.params) || 2;
+      const cur = utils.getBidIdParameter('cur', bid.params) || ['EUR'];
+      const requestPayload = {
+        id: utils.generateUUID(),
+        imp: smartxReq,
+        site: {
+          id: siteId,
+          page: page,
+          cat: cat,
+          content: 'content',
+          domain: domain,
+          publisher: {
+            id: publisherId
+          }
+        },
+        device: device,
+        at: at,
+        cur: cur
+      };
+      const userExt = {};
+      // Add GDPR flag and consent string
+      if (bidderRequest && bidderRequest.gdprConsent) {
+        userExt.consent = bidderRequest.gdprConsent.consentString;
+        if (typeof bidderRequest.gdprConsent.gdprApplies !== 'undefined') {
+          requestPayload.regs = {
+            ext: {
+              gdpr: (bidderRequest.gdprConsent.gdprApplies ? 1 : 0)
+            }
+          };
+        }
+      }
+      // Add common id if available
+      if (pubcid) {
+        userExt.fpc = pubcid;
+      }
+      // Only add the user object if it's not empty
+      if (!utils.isEmpty(userExt)) {
+        requestPayload.user = { ext: userExt };
+      }
+      // CUSTOM - Targetingparameters
+      var dataarray = [];
+      for (var i = 0; i < bid.params.user.data.length; i++) {
```

Comment:
Is this field expected to an array (or some other iterable)?  In the md file, the `data` field is shown to be an object, while the parent field `user` is shown to be an array.

When I tried to load the test params, I received a `Cannot read property 'length' of undefined` at this line, so I'm guessing something isn't quite right.  Can you please review (and possibly add a line that validates the field is properly populated before iterating over it?

Reply:
I updated the md File.